Snapchat Custom Audience destination documentation (#920)
rudderlabs/rudderstack-docs: docs/destinations/streaming-destinations/snapchat-custom-audience.mdx

<div class="infoBlock">
Find the open source code for this destination in the <a href="https://github.com/rudderlabs/rudder-transformer/tree/master/v0/destinations/">GitHub repository</a>.
</div>

## Getting started

RudderStack supports sending events to Snapchat Custom Audience via the following <Link to="/destinations/rudderstack-connection-modes/">connection modes</Link>:

| **Connection Mode** | **Web**       | **Mobile**    | **Server**    |
| :------------------ | :------------ | :------------ | :------------ |
| **Device mode**     |       -        |        -       |        -       |
| **Cloud mode**      |       -        |           -    |    Supported           |

Once you have confirmed that the source platform supports sending events to Snapchat Custom Audience, follow these steps:

1. From your [RudderStack dashboard](https://app.rudderstack.com/), add the source. Then, from the list of destinations, select **Snapchat Custom Audience**.
2. Assign a name to your destination and click **Continue**.

### Connection settings 

To successfully set up Snapchat Custom Audience as a destination, you need to configure the following settings:

> - **Connect with Snapchat Custom Audience**: Click this button to give RudderStack access to your Snapchat account.

Reviewer: Will it be better if we provide an image here?

Author: we're avoiding adding images for the quite straightforward options/connection settings.

> - **Account ID**: Enter your Snapchat account ID. For more information on getting your account ID, refer to the <a href="https://businesshelp.snapchat.com/s/article/biz-acct-id?language=en_US">Snapchat support page</a>.

Reviewer: We can remove this setting.

> - **Audience ID**: Enter your Snapchat segment ID. For more information on getting your segment ID, refer to the <Link to="#faq">FAQ</Link> section below.

Reviewer: ```suggestion
- **Segment ID**: Enter your Snapchat segment ID. For more information on getting your segment ID, refer to the <Link to="#faq">FAQ</Link> section below.
```

> - **Schema**: Select the schema (`EMAIL_SHA256`, `MOBILE_AD_ID_SHA256`, or `PHONE_SHA256`) from the dropdown.

Reviewer: ```suggestion
- **Schema**: Select the schema (`Email`, `Phone`, or `Mobile Ad Id`) from the dropdown.
```

<div class="warningBlock">

> The corresponding field (<code class="inline-code">email</code>, <code class="inline-code">mobileAdId</code>/<code class="inline-code">mobile_id</code>, or <code class="inline-code">mobile</code> must be present in your events. Otherwise, RudderStack will not send the events to Snapchat Custom Audience.)

Reviewer: ```suggestion
The corresponding field (<code class="inline-code">email</code>, <code class="inline-code">phone</code>, or  <code class="inline-code">mobileAdId</code>/<code class="inline-code">mobile_id</code> must be present in your events. Otherwise, RudderStack will not send those events to Snapchat Custom Audience.)
```

</div>

> - **Disable hashing**:  By default, RudderStack hashes the `email`, `phone`, and `mobileId` in the SHA-256 format. If enabled, RudderStack will **not** perform the hashing.

Reviewer: ```suggestion
- **Disable hashing**:  By default, RudderStack hashes the `email`, `phone`, and `mobileAdId` in the SHA-256 format. If enabled, RudderStack will **not** perform the hashing.
```

        "postalCode": "90009"
      },
        {
          "email": "alexis@example.com",
          "phone": "+1-202-555-0145",
          "country": "USA",
          "lastName": "Alexis",
          "firstName": "Keener",
          "postalCode": "90009"
        }
      ],
      "remove": [{
        "email": "jack@hotmail.com",
        "phone": "+1-202-555-0145",
        "country": "USA",
        "lastName": "Jack",
        "firstName": "Jackson ",
        "postalCode": "90009"
      }]
    }
  },
  "receivedAt": "2022-06-14T12:34:53.113Z",
  "request_ip": "35.231.104.198",
  "anonymousId": "760bf220-046f-4c2a-ad2b-3424be593740",
  "integrations": {
    "SHOPIFY": true
  },
  "originalTimestamp": "2022-06-14T12:34:53.514Z"
}
```

## Schema fields mapping

The following table details the mapping of the schema fields (as specified in the RudderStack dashboard) and the <a href="https://marketingapi.snapchat.com/docs/?&_ga=2.64649388.22530882.1663148304-88679508.1663136652#create-an-audience-segment">Snapchat Marketing API</a>.

 RudderStack schema field | Snapchat Custom Audience field |
| :-----| :--------|
| `properties.listData.add.email` | `users.data.[sha256(email)]` |

> | `properties.listData.add.phone` <br/> `properties.listData.add.mobile` | `users.data.[sha256(phone)]` |

Reviewer: ```suggestion
| `properties.listData.add.phone` <br/> `properties.listData.add.mobile` | `users.data.[sha256(phone/mobile)]` |
```

> | `properties.listData.add.mobileId` <br/> `properties.listData.add.mobileAdId` <br/> `properties.listData.add.mobile_id` | `users.data.[sha256(mobileId)]` |

Reviewer: ```suggestion
| `properties.listData.add.mobileId` <br/> `properties.listData.add.mobileAdId` <br/> `properties.listData.add.mobile_id` | `users.data.[sha256(mobileId/mobileAdId/mobile_id)]` |
```

<div class="warningBlock">
<code class="inline-code">email</code>, <code class="inline-code">mobileAdId</code>/<code class="inline-code">mobile_id</code>, or <code class="inline-code">mobile</code> must be present in your events depending on the <strong>Schema</strong> setting configured in the RudderStack dashboard. Otherwise, RudderStack will not send the events to Snapchat Custom Audience.
</div>

## FAQ

> ### Where can I find my Snapchat segment ID?

Reviewer: ```suggestion
### Where can I find my Snapchat Segment ID?
```

> To get your Snapchat segment ID, follow the steps below:

Reviewer: ```suggestion
To get your Snapchat Segment ID, follow the steps below:
```

  "integrations": {
    "SHOPIFY": true
  },
  "originalTimestamp": "2022-06-14T12:34:53.514Z"
}
```

## Schema fields mapping

The following table details the mapping of the schema fields (as specified in the RudderStack dashboard) and the <a href="https://marketingapi.snapchat.com/docs/?&_ga=2.64649388.22530882.1663148304-88679508.1663136652#create-an-audience-segment">Snapchat Marketing API</a>.

RudderStack normalizes the raw identifiers by following the <a href="https://marketingapi.snapchat.com/docs/#normalizing-hashing">Snapchat-prescribed rules</a> before SHA-256 hashing and sending them to Custom Audience.

<div class="infoBlock">
RudderStack groups the identifiers in batches with a maximum of 100000 identifiers per request. For more information, refer to the <a href="https://marketingapi.snapchat.com/docs/#adding-users">Snapchat Marketing API</a>.
</div>

| RudderStack schema field |  Custom Audience field | Notes |
| :-----| :--------| :----- |
| `properties.listData.add.email` | `users.data.[sha256(email)]` | RudderStack trims any leading and trailing whitespaces and converts all the characters into the lower case before hashing. |

> | `properties.listData.add.phone` <br/><br/> `properties.listData.add.mobile` | `users.data.[sha256(phone)]` <br/><br/> `users.data.[sha256(mobile)]` | RudderStack converts all the characters into the lower case. It does not remove any hyphens. |

Author: ```suggestion
| `properties.listData.add.phone` <br/><br/> `properties.listData.add.mobile` | `users.data.[sha256(phone)]` <br/><br/> `users.data.[sha256(mobile)]` |  RudderStack includes the country code and removes any double 0s in front of it. It also excludes any non-numeric characters like whitespaces, parentheses, `+`, or `-`. |
```

> | `properties.listData.add.mobileId` <br/><br/> `properties.listData.add.mobileAdId` <br/><br/> `properties.listData.add.mobile_id` | `users.data.[sha256(mobileId)]` <br/><br/> `users.data.[sha256(mobileAdId)]` <br/><br/> `users.data.[sha256(mobile_id)]` | RudderStack includes the country code and removes any double 0s in front of it. It also excludes any non-numeric characters like whitespaces, parentheses, `+`, or `-`. |

Author: ```suggestion
| `properties.listData.add.mobileId` <br/><br/> `properties.listData.add.mobileAdId` <br/><br/> `properties.listData.add.mobile_id` | `users.data.[sha256(mobileId)]` <br/><br/> `users.data.[sha256(mobileAdId)]` <br/><br/> `users.data.[sha256(mobile_id)]` | RudderStack converts all the characters into the lower case. It does not remove any hyphens. |
```

}
```

## Schema fields mapping

The following table details the mapping of the schema fields (as specified in the RudderStack dashboard) and the <a href="https://marketingapi.snapchat.com/docs/?&_ga=2.64649388.22530882.1663148304-88679508.1663136652#create-an-audience-segment">Snapchat Marketing API</a>.

 RudderStack schema field | Snapchat Custom Audience field |
| :-----| :--------|
| `properties.listData.add.email` | `users.data.[sha256(email)]` |
| `properties.listData.add.phone` <br/> `properties.listData.add.mobile` | `users.data.[sha256(phone | mobile)]` |
| `properties.listData.add.mobileId` <br/> `properties.listData.add.mobileAdId` <br/> `properties.listData.add.mobile_id` | `users.data.[sha256(mobileId | mobileAdId | mobile_id)]` |

<div class="warningBlock">
<code class="inline-code">email</code>, <code class="inline-code">mobileAdId</code>/<code class="inline-code">mobile_id</code>, or <code class="inline-code">mobile</code> must be present in your events depending on the <strong>Schema</strong> setting configured in the RudderStack dashboard. Otherwise, RudderStack will not send the events to Snapchat Custom Audience.
</div>

> ### Normalizing and hashing identifiers

Author: the engg. team suggested we include this section. cc @rohithkaza your thoughts on this, please?

> ### Normalizing and hashing identifiers

Reviewer: I think mentioning these points in the doc will give more clarity to our customer

> ### Normalizing and hashing identifiers

Author: Based on Sudip's suggestion - I've moved the normalization rules as notes in the mapping table and the other relevant information in the same section. Kindly check once @ItsSudip @rohithkaza

RudderStack normalizes the raw identifiers before SHA-256 hashing and sending them to Snapchat Custom Audience.

<div class="infoBlock">
RudderStack groups the identifiers in batches with a maximum of 100000 identifiers per request, as prescribed by Snapchat. For more information, refer to the <a href="https://marketingapi.snapchat.com/docs/#adding-users">Snapchat Marketing API</a>.
</div>

RudderStack adheres to the following <a href="https://marketingapi.snapchat.com/docs/#normalizing-hashing">normalization rules</a> prescribed by Snapchat:

- Normalizes email by trimming any leading and trailing whitespaces and converting all characters to lowercase before hashing.
- Normalizes the mobile advertiser ID by converting it into the lower case. RudderStack does not remove any hyphens.

> - Normalizes the phone numbers by including the country code and removing any double 0 in front of the country code. RudderStack also excludes any non-numeric characters like whitespaces, parentheses, `+`, or `-`.

Reviewer 2: Can we provide this information in tabular form along with the above mapping?

Author: this is done - please check once.

RudderStack normalizes the raw identifiers by following the <a href="https://marketingapi.snapchat.com/docs/#normalizing-hashing">Snapchat-prescribed rules</a> before SHA-256 hashing and sending them to Custom Audience.

<div class="infoBlock">
RudderStack groups the identifiers in batches with a maximum of 100000 identifiers per request. For more information, refer to the <a href="https://marketingapi.snapchat.com/docs/#adding-users">Snapchat Marketing API</a>.
</div>

| RudderStack schema field |  Custom Audience field | Notes |
| :-----| :--------| :----- |
| `properties.listData.add.email` | `users.data.[sha256(email)]` | RudderStack trims any leading and trailing whitespaces and converts all the characters into the lower case before hashing. |
| `properties.listData.add.phone` <br/> `properties.listData.add.mobile` | `users.data.[sha256(phone/mobile)]` | RudderStack converts all the characters into the lower case. It does not remove any hyphens. |

> | `properties.listData.add.mobileId` <br/> `properties.listData.add.mobileAdId` <br/> `properties.listData.add.mobile_id` | `users.data.[sha256(mobileId/mobileAdId/mobile_id)]` | RudderStack includes the country code and removes any double 0s in front of it. It also excludes any non-numeric characters like whitespaces, parentheses, `+`, or `-`. |

Reviewer 2: ```suggestion
| `properties.listData.add.phone` <br/> `properties.listData.add.mobile` | `users.data.[sha256(phone/mobile)]` | RudderStack includes the country code and removes any double 0s in front of it. It also excludes any non-numeric characters like whitespaces, parentheses, `+`, or `-`.  |
| `properties.listData.add.mobileId` <br/> `properties.listData.add.mobileAdId` <br/> `properties.listData.add.mobile_id` | `users.data.[sha256(mobileId/mobileAdId/mobile_id)]` | RudderStack converts all the characters into the lower case. It does not remove any hyphens. |
```
I think two comments are swapped.
